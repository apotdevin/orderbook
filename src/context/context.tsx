import { FC, createContext, useContext, useReducer } from 'react';

const getStep = (x: number) => {
  switch (x) {
    case 1:
      return 0.5;
    case 2:
      return 1;
    case 3:
      return 2.5;
    case 4:
      return 5;
    case 5:
      return 10;
    case 6:
      return 25;
    case 7:
      return 50;
    case 8:
      return 100;
    case 9:
      return 250;
    case 10:
      return 500;
    case 11:
      return 1000;
    case 12:
      return 2500;
    default:
      return 0.5;
  }
};

type State = {
  group: number;
  groupStep: number;
  limit: number;
  spread: number;
};

type ActionType =
  | {
      type: 'change';
      spread: number;
    }
  | { type: 'changeLimit'; limit: number }
  | { type: 'changeGroup'; group: number };

type Dispatch = (action: ActionType) => void;

export const StateContext = createContext<State | undefined>(undefined);
export const DispatchContext = createContext<Dispatch | undefined>(undefined);

const stateReducer = (state: State, action: ActionType): State => {
  switch (action.type) {
    case 'change':
      return { ...state, spread: action.spread };
    case 'changeLimit':
      return { ...state, limit: Math.min(Math.max(action.limit, 10), 100) };
    case 'changeGroup': {
      const group = Math.min(Math.max(action.group, 1), 12);
      return {
        ...state,
        group,
        groupStep: getStep(group),
      };
    }
    default:
      return state;
  }
};

const ContextProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(stateReducer, {
    spread: 0,
    group: 1,
    groupStep: getStep(1),
    limit: 15,
  });

  return (
    <DispatchContext.Provider value={dispatch}>
      <StateContext.Provider value={state}>{children}</StateContext.Provider>
    </DispatchContext.Provider>
  );
};

const useContextState = () => {
  const context = useContext(StateContext);
  if (context === undefined) {
    throw new Error('useContextState must be used within a ContextProvider');
  }
  return context;
};

const useContextDispatch = () => {
  const context = useContext(DispatchContext);
  if (context === undefined) {
    throw new Error('useContextDispatch must be used within a ContextProvider');
  }
  return context;
};

export { ContextProvider, useContextState, useContextDispatch };
