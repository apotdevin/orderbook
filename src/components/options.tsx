import { FC } from 'react';
import { Minus, Plus } from 'react-feather';
import { useContextDispatch, useContextState } from 'src/context/context';
import styled from 'styled-components';

const S = {
  title: styled.div`
    width: 100%;
    text-align: center;

    white-space: nowrap;
  `,

  row: styled.div`
    display: flex;
    justify-content: space-between;
    width: 100%;
    max-width: 300px;
    margin-bottom: 16px;
    font-size: 14px;
  `,

  column: styled.div`
    display: flex;
  `,

  button: styled.button`
    border: none;
    background: transparent;
    color: white;
    cursor: pointer;
    outline: none;
    text-decoration: none;

    :hover {
      color: black;
    }
  `,

  text: styled.div`
    width: 70px;
  `,

  buttons: styled.div`
    display: flex;
    margin-left: 8px;
  `,
};

export const Options: FC<{ spread: number; price: number }> = ({
  spread,
  price,
}) => {
  const { limit, group, groupStep } = useContextState();
  const dispatch = useContextDispatch();

  const spreadPercent = Math.round((spread / price) * 10000) / 100;

  return (
    <>
      <S.row>
        <S.column>
          <S.text>{`Group: ${groupStep}`}</S.text>
          <S.buttons>
            <S.button
              onClick={() =>
                dispatch({ type: 'changeGroup', group: group - 1 })
              }
            >
              <Minus size={14} />
            </S.button>
            <S.button
              onClick={() =>
                dispatch({ type: 'changeGroup', group: group + 1 })
              }
            >
              <Plus size={14} />
            </S.button>
          </S.buttons>
        </S.column>
        <S.column>
          <S.text>{`Limit: ${limit}`}</S.text>
          <S.buttons>
            <S.button
              onClick={() =>
                dispatch({ type: 'changeLimit', limit: limit - 5 })
              }
            >
              <Minus size={14} />
            </S.button>
            <S.button
              onClick={() =>
                dispatch({ type: 'changeLimit', limit: limit + 5 })
              }
            >
              <Plus size={14} />
            </S.button>
          </S.buttons>
        </S.column>
      </S.row>
      <S.row>
        <S.title>{`${spread} Spread (${spreadPercent.toFixed(2)}%)`}</S.title>
      </S.row>
    </>
  );
};
