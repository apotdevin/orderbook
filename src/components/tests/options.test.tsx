import { fireEvent, render } from '@testing-library/react';
import { ContextProvider } from 'src/context/context';
import { screen } from '@testing-library/dom';
import { Options } from '../options';

describe('Options', () => {
  test('matches snapshot', () => {
    const { asFragment } = render(
      <ContextProvider>
        <Options spread={20} price={200} />
      </ContextProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('shows spread and percent', () => {
    render(
      <ContextProvider>
        <Options spread={20} price={200} />
      </ContextProvider>
    );

    expect(screen.getByText('20 Spread (10.00%)')).toBeDefined();
  });

  test('change group', () => {
    render(
      <ContextProvider>
        <Options spread={20} price={200} />
      </ContextProvider>
    );

    fireEvent(
      screen.getByTestId('increaseChangeGroup'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText('Group: 1')).toBeDefined();

    fireEvent(
      screen.getByTestId('decreaseChangeGroup'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText('Group: 0.5')).toBeDefined();
  });

  test('change limit', () => {
    render(
      <ContextProvider>
        <Options spread={20} price={200} />
      </ContextProvider>
    );

    fireEvent(
      screen.getByTestId('increaseChangeLimit'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText('Limit: 20')).toBeDefined();

    fireEvent(
      screen.getByTestId('decreaseChangeLimit'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText('Limit: 15')).toBeDefined();
  });
});
