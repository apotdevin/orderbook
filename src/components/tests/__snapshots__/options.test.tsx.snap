// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="options__wrapper-sc-1wxkfq2-0 iLoTFX"
  >
    <div
      class="options__row-sc-1wxkfq2-2 sUSpG"
    >
      <div
        class="options__column-sc-1wxkfq2-3 bFEMcp"
      >
        <div
          class="options__text-sc-1wxkfq2-5 lelrqU"
        >
          Group: 0.5
        </div>
        <div
          class="options__buttons-sc-1wxkfq2-6 kXUnru"
        >
          <button
            class="options__button-sc-1wxkfq2-4 bKjZNU"
            data-testid="decreaseChangeGroup"
          >
            <svg
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <button
            class="options__button-sc-1wxkfq2-4 bKjZNU"
            data-testid="increaseChangeGroup"
          >
            <svg
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="options__column-sc-1wxkfq2-3 bFEMcp"
      >
        <div
          class="options__text-sc-1wxkfq2-5 lelrqU"
        >
          Limit: 15
        </div>
        <div
          class="options__buttons-sc-1wxkfq2-6 kXUnru"
        >
          <button
            class="options__button-sc-1wxkfq2-4 bKjZNU"
            data-testid="decreaseChangeLimit"
          >
            <svg
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <button
            class="options__button-sc-1wxkfq2-4 bKjZNU"
            data-testid="increaseChangeLimit"
          >
            <svg
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="options__row-sc-1wxkfq2-2 sUSpG"
    >
      <div
        class="options__title-sc-1wxkfq2-1 hNpzzn"
      >
        20 Spread (10.00%)
      </div>
    </div>
  </div>
</DocumentFragment>
`;
